src/mlpack/tests/layer_test.cpp

```diff
@@ -0,0 +1,277 @@
+    {
+      input.slice(i).row(j) = a.subvec(vec_idx, vec_idx + 4).t();
+      vec_idx += 5;
+    }
+  }
+  ConvLayer<> conv3(4, 1, 3, 3, 1, 1, 1, 1);
+  arma::mat id3 = arma::zeros<arma::mat>(3, 3);
+  id3(1, 1) = 1;  
+  arma::cube conv3_w(3, 3, 4 * 1);
+  for (size_t i = 0; i < conv3_w.n_slices; ++i) 
+    conv3_w.slice(i) = id3;
+  conv3.Weights() = conv3_w;
+  
+  arma::cube output(5, 5, 1);
+  output.slice(0) = input.slice(0) * 4;
+  arma::cube convOutput;
+  conv3.InputParameter() = input;
+  conv3.Forward(conv3.InputParameter(), conv3.OutputParameter());
+  Test(conv3.OutputParameter(), output);
+  
+  arma::cube error = arma::ones(5, 5, 1);
+  conv3.Backward(conv3.InputParameter(), error, conv3.Delta());
+  Test(conv3.Delta(), arma::ones(5, 5, 4));
+
+  arma::cube delta = arma::zeros(5, 5, 1);
+  delta(2, 2, 0) = 1;
+  conv3.Gradient(conv3.InputParameter(), delta, conv3.Gradient());
+  arma::cube gradOut(3, 3, 4);
+  arma::mat grad_w;
+  grad_w << 7 << 8 << 9 << arma::endr
+        << 12 << 13 << 14 << arma::endr
+        << 17 << 18 << 19;
+  for (size_t i = 0; i < gradOut.n_slices; ++i)
+    gradOut.slice(i) = grad_w;
+  Test(conv3.Gradient(), gradOut);
+}
+
+void ConcatLayerTest()
+{
+  arma::cube input(5, 5, 3, arma::fill::ones);
+  ConvLayer<> convLayer0(3, 2, 1, 1);
+  ConvLayer<> convLayer1(3, 3, 1, 1);
+  auto JoinLayers = std::tie(convLayer0, convLayer1);
+  size_t numLayers = 2;
+  ConcatLayer<decltype(JoinLayers), arma::cube, arma::cube> 
+      concatLayer0(numLayers, std::move(JoinLayers));
+
+  arma::mat id = arma::ones(1, 1);
+
+  arma::cube convLayer0_w(1, 1, 3 * 2);
+  for (size_t i = 0; i < convLayer0_w.n_slices; ++i)
+    convLayer0_w.slice(i) = id;
+  convLayer0.Weights() = convLayer0_w;
+
+  arma::cube convLayer1_w(1, 1, 3 * 3);
+  id(0, 0) = 2;
+  for (size_t i = 0; i < convLayer1_w.n_slices; ++i)
+    convLayer1_w.slice(i) = id;
+  convLayer1.Weights() = convLayer1_w;
+
+  arma::cube d1, d2; // dummy cubes.
+  arma::cube output;
+
+  //! Forward pass test for ConcatLayer...
+  convLayer0.InputParameter() = convLayer1.InputParameter() = input;
+  convLayer0.Forward(convLayer0.InputParameter(), convLayer0.OutputParameter());
+  convLayer1.Forward(convLayer1.InputParameter(), convLayer1.OutputParameter());
+  concatLayer0.Forward(d1, output);
+  arma::cube desiredOutput = arma::join_slices(convLayer0.OutputParameter(), 
+                              convLayer1.OutputParameter());
+  Test(output, desiredOutput);
+  
+  //! Backward pass test for ConcatLayer.
+  arma::cube error = arma::ones(5, 5, 5);
+  concatLayer0.Backward(d1, error, d2);
+  arma::cube backout0 = arma::ones(5, 5, 3) * 2;
+  Test(convLayer0.Delta(), backout0);
+  arma::cube backout1 = arma::ones(5, 5, 3) * 6;
+  Test(convLayer1.Delta(), backout1);
+
+  //! todo: Gradient update test for ConcatLayer.
+}
+
+
+BOOST_AUTO_TEST_CASE(SampleInceptionLayerTest)
+{
+  SampleInceptionNetwork();
+  SmallNetworkTest();
+  ConvLayerTest();
+  ConcatLayerTest();
```

Comment:
Maybe you can consider put these test cases in different test case, like

```
BOOST_AUTO_TEST_CASE(SampleInceptionNetworkTest)
{
}
```

This way the error messages of test case should be better